lib/libzfs_core/libzfs_core.c

```diff
@@ -586,6 +589,116 @@ lzc_get_holds(const char *snapname, nvlist_t **holdsp)
 	return (lzc_ioctl(ZFS_IOC_GET_HOLDS, snapname, NULL, holdsp));
 }
 
+static unsigned int
+max_pipe_buffer(int infd)
+{
+#if __linux__
+	static unsigned int max;
+	if (max == 0) {
+		max = 1048576; /* fs/pipe.c default */
+
+		FILE *procf = fopen("/proc/sys/fs/pipe-max-size", "re");
+		if (procf != NULL) {
+			if (fscanf(procf, "%u", &max) <= 0) {
+				/* ignore error: max untouched if parse fails */
+			}
+			fclose(procf);
+		}
+	}
+
+	unsigned int cur = fcntl(infd, F_GETPIPE_SZ);
+	if (cur < max && fcntl(infd, F_SETPIPE_SZ, max) != -1)
+		cur = max;
+	return (cur);
+#else
+	/* FreeBSD automatically resizes */
+	(void) infd;
+	return (BIG_PIPE_SIZE);
+#endif
+}
+
+#if __linux__
+struct send_worker_ctx {
+	int from;	/* read end of pipe, with send data; closed on exit */
+	int to;		/* original arbitrary output fd; mustn't be a pipe */
+};
+
+static void *
+send_worker(void *arg)
+{
+	struct send_worker_ctx *ctx = arg;
+	unsigned int bufsiz = max_pipe_buffer(ctx->from);
+	ssize_t rd;
+
+	while ((rd = splice(ctx->from, NULL, ctx->to, NULL, bufsiz,
+	    SPLICE_F_MOVE | SPLICE_F_MORE)) > 0)
+		;
+
+	int err = (rd == -1) ? errno : 0;
+	close(ctx->from);
+	return ((void *)(uintptr_t)err);
+}
+#endif
+
+/*
+ * Since Linux 5.10, 4d03e3cc59828c82ee89ea6e27a2f3cdf95aaadf
+ * ("fs: don't allow kernel reads and writes without iter ops"),
+ * ZFS_IOC_SEND* will EINVAL when writing to /dev/null, /dev/zero, &c.
+ *
+ * This wrapper transparently executes func() with a pipe
+ * by spawning a thread to copy from that pipe to the original output
+ * in the background.
+ *
+ * Returns the error from func(), if nonzero,
+ * otherwise the error from the thread.
+ *
+ * No-op if orig_fd is -1, already a pipe (but the buffer size is bumped),
+ * and on not-Linux; as such, it is safe to wrap/call wrapped functions
+ * in a wrapped context.
+ */
+int
+lzc_send_wrapper(int (*func)(int, void *), int orig_fd, void *data)
+{
+#if __linux__
+	struct stat sb;
+	if (orig_fd != -1 && fstat(orig_fd, &sb) == -1)
+		return (errno);
+	if (orig_fd == -1 || S_ISFIFO(sb.st_mode)) {
+		if (orig_fd != -1)
+			(void) max_pipe_buffer(orig_fd);
+		return (func(orig_fd, data));
+	}
+	if ((fcntl(orig_fd, F_GETFL) & O_ACCMODE) == O_RDONLY)
+		return (errno = EBADF);
```

Comment:
No objection really, but I'm curious what callers are depending on `errno` to be set instead of depending the return value?

Reply:
hell if I know (didn't look, doesn't help that errno is an annoyingly persistent global), but all extant code paths set errno too (or return errno), and Rich's original apparently had Some callers yielding "error: [message]: Success", so.

if this weren't lzc and we hadn't already provided return==errno before I'd probably say to hell with it, but; meh